File: src/cpp_basic_nodes/src/simple_publisher.cpp
/*
 * This program does the following
 *      1. Demonstrates how to create a publisher using the NodeHandle
 *      2. Introduces executing things at a frequency
 */

// Include ROS header file
#include <ros/ros.h>
// Include the header file for the message content
/*
 * Every message actually has a data structure to it (implemented as a class in C++)
 * This class is stored in a header file which you must include in order to use that message
 * 
 * Location on local PC: /opt/ros/noetic/include, inside std_msgs/String.h
 */
#include <std_msgs/String.h>
// Other header files
#include <iostream>

using namespace std;

int main(int argc, char **argv) {
    // Initialize the node
    ros::init(argc, argv, "simple_cpp_publisher");
    
    // Initialize the NodeHandle
    /*
     * The first parameter is the namespace parameter. The ~ character makes the names of all the topics 
     * and services prefixed with /node_name/ (which is /simple_cpp_publisher/ here)
     * 
     * More here: http://wiki.ros.org/Names
     * Constructor reference: http://docs.ros.org/en/latest/api/roscpp/html/classros_1_1NodeHandle.html#a0eae6a7fd8c216c3f0860aec44ca81ba
     */
    ros::NodeHandle nh("~");
    
    // Create a publisher object
    /*
     * We will use this object to publish messages
     * The template parameter (in <>) is the class of the message type
     * The first parameter is the topic name and the second parameter is the queue size
     * 
     * Link: http://docs.ros.org/en/latest/api/roscpp/html/classros_1_1NodeHandle.html#a6b655c04f32c4c967d49799ff9312ac6
     */
    ros::Publisher str_publisher = nh.advertise<std_msgs::String>("hello_str", 1000);
    
    // Create a rate controller
    /*
     * For the purpose of this tutorial, the messages will be published in a loop, once every two seconds (i.e. 0.5 Hz)
     * 
     * Link: http://docs.ros.org/en/latest/api/rostime/html/classros_1_1Rate.html#ad7ef59c5fb4edb69c6a9471987c3117d
     */
    ros::Rate rate_ctrl(0.5);
    
    // Some variables to keep track of things
    long int seq = 0;   // Sequence number
    
    // Loop while ROS is "ok"
    /*
     * ros::ok() is false when shutdown is called
     * 
     * Link: http://docs.ros.org/en/latest/api/roscpp/html/namespaceros.html#a276d68870be2125b1cde229fee013e45
     */
    while (ros::ok()) {
        // Construct a message
        /*
         * This object will be assigned all data and given to the ros::Publisher object created earlier.
         * The underlying ROS infrastructure will then take the message and publish it to the topic.
         */
        std_msgs::String msg;
        /* 
         * A string object which will finally be assigned to the message
         * Reference: http://www.cplusplus.com/reference/string/
         */
        std::string str_msg = "Hello, World!";
        str_msg += " [I: " + std::to_string(seq) + "]"; // Append sequence number
        seq += 1;   // Increment to the next sequence
        msg.data = str_msg;    // Assign it data
        
        // Publish it on the topic
        /*
         * Use the publish member function to publish a message
         * 
         * Link: http://docs.ros.org/en/latest/api/roscpp/html/classros_1_1Publisher.html#a0933b8fb71fa577e0eeed771d37631ed
         */
        str_publisher.publish(msg);
        
        // Print it as INFO message
        ROS_INFO_STREAM("Published " << str_msg);   // You can verify the time and sequence number for the rate (during execution)
        
        // Call the ros::spinOnce function
        /*
         * This is ideal to the ros::spin(), but this will return control to the loop and the execution will resume from after the call point
         * 
         */
        ros::spinOnce();
        
        // Cause a delay to obtain the frequency
        /*
         * Causes a delay. Since we gave a frequency of 0.5 Hz, the delay will be of approx. 2 seconds
         * 
         * Link: http://docs.ros.org/en/latest/api/rostime/html/classros_1_1Rate.html#ae5664d27cda1b17a103347560259e945
         */
        rate_ctrl.sleep();
    }
}

File: src/py_basic_nodes/scripts/simple_publisher.py
#!/usr/bin/env python3

"""
    Main objective of this program is to do the following:
        1. Demonstrate how to create a publisher
        2. Introduces executing things at a particular frequency
"""

import rospy
import sys
# Import the library generated for the message file
"""
    Every message actually has a data structure to it (implemented as a class in Python)
    This class is stored in a file which you must import in order to use the class

    Location on local PC: /opt/ros/noetic/lib/python3/dist-packages, inside std_msgs/msg/_String.py
"""
from std_msgs.msg import String

def main():
    # Initialize node
    rospy.init_node("simple_py_publisher", argv=sys.argv, anonymous=False)

    # Create a publisher object
    """
        We will use this object to publish messages
        First argument is the topic name, second argument is the class (message type) and third is the queue size
        The first argument uses the node name for the topic name (to get it under a namespace)

        Link (for names in ROS): http://wiki.ros.org/Names
        Link (for Publisher class): http://docs.ros.org/en/melodic/api/rospy/html/rospy.topics.Publisher-class.html
        Link (for get_name()): http://docs.ros.org/en/melodic/api/rospy/html/rospy-module.html#get_name
    """
    pub_obj = rospy.Publisher("{0}/hello_str".format(rospy.get_name()), String, queue_size=1000)

    # Create a rate controller
    """
        This is used to cause a delay to suit a particular frequency or duration
        Here, we will publish once every two seconds (i.e. frequency is 0.5 Hz)

        Link: http://docs.ros.org/en/melodic/api/rospy/html/rospy.timer.Rate-class.html
    """
    rt_ctrl = rospy.Rate(0.5)

    # Some variables to keep track of things
    seq = 0     # Sequence number of the message number

    # Loop execution until node hasn't shutdown
    while not rospy.is_shutdown():
        # Create a message object
        """
            This message will be finally published in the end
        """
        msg_obj = String()
        msg_obj.data = "Hello, World! [I: {}]".format(seq)  # Fill data
        seq += 1    # Increment sequence number

        # Publish the message
        """
            Use the publish member function to publish a message

            Link: http://docs.ros.org/en/melodic/api/rospy/html/rospy.topics.Publisher-class.html#publish
        """
        pub_obj.publish(msg_obj)

        # Log the message published
        rospy.loginfo("Published '{}'".format(msg_obj.data))

        # Cause a delay to obtain the frequency
        """
            Create a delay using the sleep function. This will hold the execution according to the rate.

            Link: http://docs.ros.org/en/melodic/api/rospy/html/rospy.timer.Rate-class.html#sleep
        """
        rt_ctrl.sleep()
        pass
    pass

if __name__ == "__main__":
    try:
        main()
    except rospy.ROSException as e:
        rospy.logfatal("A ROS Exception has occurred: {0}".format(e))
